core/engine/src/value/inner/nan_boxed.rs

```diff
@@ -551,103 +449,128 @@ impl NanBoxedValue {
         Self::from_inner_unchecked(bits::tag_bool(value))
     }
 
-    /// Returns a `InnerValue` from a boxed `[JsBigInt]`.
+    /// Returns a `InnerValue` from a boxed [`JsBigInt`].
     #[must_use]
     #[inline(always)]
     pub(crate) fn bigint(value: JsBigInt) -> Self {
-        Self::from_inner_unchecked(unsafe { bits::tag_bigint(Box::new(value)) })
+        let value = value.into_raw().cast::<()>().cast_mut();
+        let addr = bits::tag_bigint(value);
+        Self {
```

Comment:
Adding a function like:

```rust
#[must_use]
#[inline(always)]
fn new_unchecked(ptr: u64) -> Self {
  // ...
}
```

would greatly simplify all those constructors.

Reply:
That doesn't work. It wouldn't preserve the provenance of the original pointer. However, something like this should work:
```rust
fn new_pointer(ptr: *mut (), addr: usize) -> Self {
    // ...
}
```